qemu/tests/hotplug_mem_share_discard_data.py

```diff
@@ -0,0 +1,93 @@
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    timeout = float(params.get("login_timeout", 240))
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=timeout)
+
+    mem_name = params["target_mems"]
+    hp_size = utils_memory.read_from_meminfo("Hugepagesize")
+    hp_free = int(utils_memory.read_from_meminfo('HugePages_Free'))
+    size_target_mem = params["size_mem_%s" % mem_name]
+
+    hp_total = int(utils_memory.read_from_meminfo("HugePages_Total"))
+    hp_plugged = int(normalize_data_size(size_target_mem, "K"))
+    hp_target = int(hp_plugged / hp_size)\
+        + int(hp_total)
+    process.system("echo %s > /proc/sys/vm/nr_hugepages" % hp_target, shell=True)
+
+    is_discard_data = params.get("discard-data_plug1", "yes")
+    is_backend_file = params["backend_mem_plug1"]
+
+    hotplug_test = MemoryHotplugTest(test, params, env)
+    hotplug_test.hotplug_memory(vm, mem_name)
+    hotplug_test.check_memory(vm)
+
+    # Checks if hotplugged device has share: True option
+    if utils_misc.compare_qemu_version(6, 1, 0, is_rhev=False) is True:
+        mem_dev_name = "mem-%s" % params["target_mems"]
+        query_list = vm.monitor.info("memdev")
+        for dev in query_list:
+            error_context.context("device id: %s has share: %s" % (dev['id'], dev['share']), test.log.debug)
+            if dev['id'] == mem_dev_name and dev['share'] is not True:
+                raise exceptions.TestFail("dev %s has share option: %s" % (dev['id'], dev['share']))
+
+    hotplug_test.unplug_memory(vm, mem_name)
+    hotplug_test.check_memory(vm)
+
+    session.close()
+    error_context.context("Shutdown guest...")
+    vm.destroy()
+
+    hp_total = int(utils_memory.read_from_meminfo("HugePages_Total"))
+    hp_free = int(utils_memory.read_from_meminfo('HugePages_Free'))
+
+    error_context.context("hp_total: %s" % str(hp_total), test.log.debug)
+    error_context.context("hp_free: %s" % str(hp_free), test.log.debug)
+
+    if is_backend_file == "memory-backend-file":
+        if is_discard_data == "no":
+            try:
+                process.system("ls %s" % params["mem-path_plug1"])
+            except process.CmdError:
+                raise exceptions.TestError("Error, %s not found." % params["mem-path_plug1"])
+
+            op = (hp_total - hp_free) * (hp_size / 1024)
+            hp_used = int(normalize_data_size("%sM" % str(op), "K"))
+
+            error_context.context("hp_used: %s" % str(hp_used), test.log.debug)
+
+            if hp_used != hp_plugged:
+                raise exceptions.TestError("Error, total hugepages doesn't match with used memory")
+        elif hp_total != hp_free:
+            raise exceptions.TestError("Error, free hugepages doesn't match with total hugepages")
+        # Deletes the mem-path file to avoid test error
+        process.system("rm -rf %s" % params["mem-path_plug1"])
```

Comment:
After mem-path_plug1 be deleted,  there is another step to confirm the mem be releaed.

Reply:
Changed!